### PR TITLE
introduce fact "ansible_processor_nproc": number of usable vcpus

### LINKED ISSUE #51504: ansible_processor_vcpus fact incorrect for containers
##### SUMMARY
As mentioned in https://github.com/ansible/ansible/issues/2492, gathering facts for a machine using OpenVZ, LXC and possibly other container systems produces the wrong number of CPU cores:

**ansible -m setup *hostname***

```
"ansible_processor": [
  "ansible_processor_cores": 8,
  "ansible_processor_count": 1,
  "ansible_processor_threads_per_core": 2,
  "ansible_processor_vcpus": 16,
]
```

This is because ansible is not correctly determining the true number of cores available to the system, which is accurately determined via `grep -c ^processor /proc/cpuinfo`, `nproc` and other commands.

https://github.com/ansible/ansible/issues/2492 was closed because it was decided this was the "correct" behavior, but I assure you it is not. Rarely does a script need to know the total hardware cores vs the actual available cores it can utilize.

For a real word example, setting the number of worker_processes in an Nginx configuration by using `ansible_processor_vcpus` will greatly impact performance if we set it to 16 vs 4. 

The workaround is defining a task which registers a variable using a shell command like the ones mentioned above. But what if we need to include this behavior in multiple roles? It's only a single task so factoring it out into a separate role is extreme overkill. This particular fact is important enough that it should be readily available to the user as a machine fact.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
setup

##### ANSIBLE VERSION
```
ansible 2.7.5
  config file = /home/user/.ansible.cfg
  configured module search path = ['/home/user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.2 (default, Jan 16 2019, 19:49:22) [GCC 8.2.1 20181215 (Red Hat 8.2.1-6)]

```

##### CONFIGURATION
Stock configuration.

##### OS / ENVIRONMENT
CentOS 7, OpenVZ / Virtuozzo container

##### STEPS TO REPRODUCE
Run `ansible -m setup <hostname>`

##### EXPECTED RESULTS
```
"ansible_processor": [
  "ansible_processor_cores": 4,
  "ansible_processor_count": 1,
  "ansible_processor_threads_per_core": 1,
  "ansible_processor_vcpus": 4,
]
```

##### ACTUAL RESULTS
```
"ansible_processor": [
  "ansible_processor_cores": 8,
  "ansible_processor_count": 1,
  "ansible_processor_threads_per_core": 2,
  "ansible_processor_vcpus": 16,
]
```

### PR DESCRIPTION
**Pull request by lpirl**
##### SUMMARY
E.g. in containers with limits applied, ``ansible_processor_vcpus`` might not reflect the number of actually visible/usable "vcpus".
This PR introduces ``ansible_processor_nproc``, which is meant to report the same as the command ``nproc`` (which "print[s] the number of processing units available") and thus reflect the number of actually visible/usable "vcpus".

Fixes #51504.

##### ISSUE TYPE
- Bugfix / Feature Pull Request

##### COMPONENT NAME
(Linux hardware) facts

##### ADDITIONAL INFORMATION
By the way, due to
https://github.com/ansible/ansible/blob/b59fa134be844a69990fa1b384f7470013b009ef/lib/ansible/module_utils/facts/hardware/linux.py#L188
this can never be true:
https://github.com/ansible/ansible/blob/b59fa134be844a69990fa1b384f7470013b009ef/lib/ansible/module_utils/facts/hardware/linux.py#L211
, no? So line 211-212 could be removed, I guess.

### CODE REVIEW

**Review by samdoran — COMMENTED** [review]
Overall, this change looks good. I think I would like to see some more fixtures that result in `nproc` differing from `cores` but we have a couple.

This is a nice simple change using existing information. Is there a chance that the output of `nproc` would differ from this? I'm wondering if we could try and use the `nproc` output but fall back to using `processor_occurence` if `nproc` isn't available. It may not be worth the added complexity, though.

Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example. This will be `minor_changes` rather than `bugfixes`.

It may also be a good idea to write a [porting guide entry](https://github.com/ansible/ansible/blob/e6ca730c5e5698662962e14795c790e6cb5f2c95/docs/docsite/rst/porting_guides/porting_guide_2.10.rst) explaining this new fact is available. This has been something folks have been wanting/needing for a while, so calling it out in a porting guide would highlight its availability.

**Comment by lpirl** [comment]
Thanks for the feedback @samdoran.

> Is there a chance that the output of nproc would differ from this?

When creating the PR, I was asking this myself but I didn't actually went any further in finding out.

Now, after taking a peek at the source of coreutils' `nproc`, I can tell the following: `nproc` does much more than – loosely speaking – `egrep -c '^processor\s:.*$' /proc/cpuinfo`.
For example, `/proc` might not be mounted. `nproc` prefers the use of system calls to determine the number of available processors.
Another circumstance is, that the number of processors available to a thread *might* be smaller than the number of available processors available to the scheduler, which in turn *might* be smaller than the number of processors configured in the system.
You can imagine that there are more corner cases and once again, we are surprised that a tasks appearing very simple can get very complex…
See coreutils' `lib/nproc.c` for all the details.

> I'm wondering if we could try and use the nproc output but fall back to using processor_occurence if nproc isn't available. It may not be worth the added complexity, though.

After having looked into the source, getting an idea about corner cases and considering the variable is named `…_nproc`, maybe we should stick to the output of `nproc` all the way? In the unlikely case that coreutils are not installed, it would be okay for me (as a user) that this variable would be undefined.

So, what you think?

**Comment by samdoran** [comment]
> Now, after taking a peek at the source of coreutils' nproc, I can tell the following: nproc does much more than – loosely speaking – egrep -c '^processor\s:.*$' /proc/cpuinfo.

I did the same thing, which is why I asked. There's a lot going on in there and it'd be nice to leverage this tool _if_ it's available. That's the tricky part.

>  In the unlikely case that coreutils are not installed, it would be okay for me (as a user) that this variable would be undefined.

The problem with facts being inconsistently defined is it can result it undefined errors in playbooks and making it hard to rely on those facts. I think it's better to have it always defined rather only defined if successful.

**Comment by lpirl** [comment]
Thanks @samdoran.
I added a changelog fragment and an entry to the porting guide. 

As a proposal, the implementation to get the number vcpus available to the (remote) process will now try the following (in order):
* on Python >= 3: use standard library,
* on Python < 3: call ``nproc`` 
* if calling ``nproc`` fails, fall back to what ``/proc`` reports (which is the number of vcpus available to the scheduler, I guess)

Additionally, I thought about naming the fact ``ansible_processor_vcpus_available`` to make the semantics a bit more transparent.

Opinions welcome.

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 1]
```diff
@@ -275,6 +275,21 @@ def get_cpu_facts(self, collected_facts=None):
                 cpu_facts['processor_vcpus'] = (cpu_facts['processor_threads_per_core'] *
                                                 cpu_facts['processor_count'] * cpu_facts['processor_cores'])
 
+                if sys.version_info[0] >= 3:
```
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/linux.py]: Import `PY3` from `six`
```suggestion
                if PY3:
```

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 2]
```diff
@@ -275,6 +275,21 @@ def get_cpu_facts(self, collected_facts=None):
...
+                    cpu_facts['processor_nproc'] = len(
+                        os.sched_getaffinity(0)
```
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/linux.py:286]: It's great that Python has this capability, but it's not always avaialble. In this conext (Python >=3 on Linux systems) it _should_ be available, but it's probably best to catch `AttributeError` just in case it's not available.

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 3]
```diff
@@ -275,6 +275,21 @@ def get_cpu_facts(self, collected_facts=None):
...
+                    )
+                else:
+                    cmd = ('nproc',)
+                    rc, out, _ = self.module.run_command(cmd)
```
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/linux.py]: This needs to go through `get_bin_path()` in order to get the absolute path to the executable since the execution `PATH` in the remote environment is far from consistent.
```suggestion
                   try:
                       cmd = get_bin_path('nproc')
                   except ValueError:
                       cpu_facts['processor_nproc'] = processor_occurence
                       
                    rc, out, err = self.module.run_command(cmd)
```

**Review by samdoran — CHANGES_REQUESTED** [review]
>Additionally, I thought about naming the fact ansible_processor_vcpus_available to make the semantics a bit more transparent.

Hmmm, I think that's a bit too wordy. Maybe just `ansible_processor_available`.

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 3 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py]: very good, thanks

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 2 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: Hm, I'd find it odd to not trust in what the docs of the standard library say. The docs do not state platform-dependent availability of the method.
But I totally see your point and checked when this method has been introduced. In 3.3. ([search in 3.2.6 docs with no results](https://docs.python.org/release/3.2.6/search.html?q=affinity), [search in 3.3.0 docs with results](https://docs.python.org/release/3.3.0/search.html?q=affinity)) and made the version check more accurate. Hence, I couldn't use ``PY3`` or is there another way to check for minor version numbers using ``six``?

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 1 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py]: see below

**Comment by lpirl** [comment]
> > Additionally, I thought about naming the fact ansible_processor_vcpus_available to make the semantics a bit more transparent.
> 
> Hmmm, I think that's a bit too wordy. Maybe just `ansible_processor_available`.

That'd be too unspecific for my taste – let's stick to `…_nproc` then. :)

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: > The docs do not state platform-dependent availability of the method.

In particular, this isn't available on macOS. I found this caveat in the Python docs [here](https://docs.python.org/3/library/os.html#interface-to-the-scheduler):
> They are only available on some Unix platforms.

Again, probably not a big deal since Ansible should only be using this file on Linux where it should be available. But I like to be overly cautious.

> ...introduced. In 3.3. ... and made the version check more accurate.

Ansible only supports Python >= 3.5, so testing for `PY3` is sufficient.

File: docs/docsite/rst/porting_guides/porting_guide_2.10.rst [hunk 4]
```diff
@@ -115,6 +115,9 @@ Noteworthy plugin changes
 -------------------------
 
 * The ``hashi_vault`` lookup plugin now returns the latest version when using the KV v2 secrets engine. Previously, it returned all versions of the secret which required additional steps to extract and filter the desired version.
+* The ``setup`` module now provides ``ansible_processor_nproc`` which
+  reflects the number of vcpus available to processes (falls back to the
+  number of vcpus available to the scheduler)
```
Comment (samdoran) [inline docs/docsite/rst/porting_guides/porting_guide_2.10.rst]: This isn't a plugin change (fact gathering is really an internal implementation). Since this is a change to the default gathered facts, put this under the `Playbook` section. It can just be its own paragraph.

I would omit the mention of ``setup`` and just describe the new fact.
```suggestion
A new fact, ``ansible_processor_nproc``, reflects the number of vcpus available. This is useful in containers or other environments where limits are applied and the CPUs available on the host are not the same as CPUs available to the container. 
```

Comment (lpirl) [reply docs/docsite/rst/porting_guides/porting_guide_2.10.rst]: Proposal pushed.

**Comment by lpirl** [comment]
If we defensively catch the ``AttributeError`` anyway (see diff), we *could* omit checking for ``PY3`` (as in the diff at the moment).
Pro: less nesting, higher readability; con: less explicit, exceptions are expensive.

Also, there is a new proposal for the change to the porting guide.

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: I think this approach is good. Exceptions in Python are not as expensive as in other languages so I don't see a problem with this approach. I would suggest reworking this a bit to make it easier for the reader, namely returning early when possible and eliminating the need for a conditional.

```python
                # Default value unless we find something better
                cpu_facts['processor_nproc'] = processor_occurence

                try:
                     cpu_facts['processor_nproc'] = len(
                         os.sched_getaffinity(0)
                     )
                except AttributeError:
                    # in Python < 3.3 and MacOS (shouldn't happen here,
                    # but to be safe) ``sched_getaffinity`` is not available
                    try:
                        cmd = get_bin_path('nproc')
                    except ValueError:
                        return cpu_facts

                    rc, out, _err = self.module.run_command(cmd)
                    if rc == 0:
                        cpu_facts['processor_nproc'] = int(out)
```

The `cpu_facts.setdefault()` line can be omitted with this approach. This may be more lines of code, but I think it's a bit easier to follow. Just my opinion.

Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: Moving the assignment of the default value up is fine with me.
"Replacing" the `else` of the `try … except` with a `return` wouldn't increase readability for me.
Also, if code needs to be added below the computation of `processor_nproc`, it is probably easier if we do not return in corner cases but keep going.
But if you insist, I wouldn't veto.

Comment (samdoran) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: Ok, let's move the default assignment up and keep the `else` without returning early. The main way I think this should read is "start with `processor_occurence` and use something better if we can".

Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: So just as the PR is right now, no?

**Comment by lpirl** [comment]
Are we ready to merge this @samdoran ?

File: lib/ansible/module_utils/facts/hardware/linux.py [hunk 5]
```diff
@@ -275,6 +275,27 @@ def get_cpu_facts(self, collected_facts=None):
...
+
+                try:
+                    cpu_facts['processor_nproc'] = len(
+                        os.sched_getaffinity(0)
+                    )
+                except AttributeError:
+                    # in Python < 3.3 and MacOS (shouldn't happen here,
+                    # but to be safe) ``sched_getaffinity`` is not available
+                    try:
+                        cmd = self.module.get_bin_path('nproc')
```
Comment (samdoran) [inline lib/ansible/module_utils/facts/hardware/linux.py]: I moved `get_bin_path` out of `AnsibleModule` and into a standalone function in `lib/ansible/module_utils/common/process.py`. I would recommend using that function directly rather than the method since that is what I would like to move the rest of the code to using eventually.

**Comment by samdoran** [comment]
There are a lot of unit test failures in the latest run.

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 5 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py]: Fixed.

**Comment by lpirl** [comment]
> There are a lot of unit test failures in the latest run.

The tests that failed in the most recent run seem to fail because the whole mocking does not work when we call `nproc` (e.g.: the mocks expects an ARM single vcpu system but the test runs on a dual vcpu system, which leads to the assertion `1 == 2` which makes the test fail).

Fixing the test infrastructure for the CPU facts is another cup of tea, I guess…

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 2 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py:286]: > So just as the PR is right now, no?

Is this resolved? Any obligations left for a merge?

File: lib/ansible/module_utils/facts/hardware/linux.py (hunk 5 above)
Comment (lpirl) [reply lib/ansible/module_utils/facts/hardware/linux.py]: I am going to resolve this on my own to raise attention again…

**Comment by lpirl** [comment]
I'd be very happy if we could land this @samdoran . We addressed all concerns in the reviews and it is a bit of work to rebase this PR on current devel now and then.

**Comment by samdoran** [comment]
I've been away on paternity leave. The changes look good we just need to fix the unit tests. Maybe we need to add more mocks/fixtures? I'll take a look and see if I can come up with something.

**Comment by samdoran** [comment]
I'm making progress adding additional mocks and fixture data.

**Comment by lpirl** [comment]
Alright, thanks for taking care – glad you dig in.
(And congrats ;))

**Comment by samdoran** [comment]
I got the tests passing locally. We'll see what Shippable says. Supporting Python 2.6 continues to be a fun exercise. 😄

**Comment by lpirl** [comment]
> I got the tests passing locally. We'll see what Shippable says. Supporting Python 2.6 continues to be a fun exercise.

Heroic! :)

**Comment by lpirl** [comment]
Sweet, thanks so much @samdoran .

**Comment by samdoran** [comment]
No problem. Thanks for the contribution.
